src/main/java/com/artozersky/HackerNewsAPI/service/impl/NewsPostServiceImpl.java

```diff
@@ -0,0 +1,262 @@
+        //cacheService.putPostInCache(postModel);
+        //should i update the cache with the top posts???
+
+        return convertToDTO(postModel);
+    }
+
+    @Override
+    public List<NewsPostsResponseDTOImpl> getAllPosts(int limit) {
+
+        if (limit <= 0) {
+            throw new CustomServiceException("Limit must be greater than 0");
+        }
+
+        Pageable pageable = PageRequest.of(0, limit);
+        Page<NewsPostModelImpl> page = postRepository.findAll(pageable);
+
+        return page.getContent().stream()
+                .map(this::convertToDTO)
+                .collect(Collectors.toList());
+    }
+
+    @Override
+    public List<NewsPostsResponseDTOImpl> getTopPosts(int limit) {
+
+    List<NewsPostModelImpl> cachedPosts = cacheService.getAllPostsFromCache();//stale will be checked inside
+
+    int postsNeeded = limit - cachedPosts.size();
+
+    System.out.println("cached posts size: " + cachedPosts.size() );
+
+    System.out.println("posts needed: " + postsNeeded );
+
+    List<NewsPostModelImpl> dbPosts = new ArrayList<>();
+
+    dbPosts = fetchPostsFromDbIfNeeded(postsNeeded, cachedPosts);
+
+    System.out.println("fetched from db: " + dbPosts.size() );
+
+    List<NewsPostModelImpl> combinedPosts = new ArrayList<>(cachedPosts);
+    combinedPosts.addAll(dbPosts);
+
+    return combinedPosts.stream()
+        .map(this::convertToDTO)
+        .collect(Collectors.toList());
+}
+
+    
+    @Override
+    public NewsPostsResponseDTOImpl savePost(@Valid NewsPostsCreateDTOImpl postCreateDTO) {
+        
+        NewsPostModelImpl post = modelMapper.map(postCreateDTO, NewsPostModelImpl.class);
+        
+        post.initialize();
+        
+        NewsPostModelImpl savedPost = postRepository.save(post);
+        
+        NewsPostsResponseDTOImpl responseDTO = modelMapper.map(savedPost, NewsPostsResponseDTOImpl.class);
+        
+        responseDTO.setMessage("Post created successfully");
+
+        updateCacheWithTopPosts();
+        
+        return responseDTO;
+    }
+
+    @Override
+    public NewsPostsResponseDTOImpl updatePost(NewsPostsUpdateDTOImpl postUpdateDTO, Long postId) {
+        
+        NewsPostModelImpl post = postRepository.findById(postId)
+        .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + postId));
+        
+        boolean isRedundant = true;
+        
+        if (postUpdateDTO.getTitle() != null && !postUpdateDTO.getTitle().trim().isEmpty()) {
+            if (!postUpdateDTO.getTitle().equals(post.getTitle())) {
+                post.setTitle(postUpdateDTO.getTitle());
+                isRedundant = false;
+            }
+        }
+        
+        if (postUpdateDTO.getUrl() != null && !postUpdateDTO.getUrl().trim().isEmpty()) {
+            if (!postUpdateDTO.getUrl().equals(post.getUrl())) {
+                post.setUrl(postUpdateDTO.getUrl());
+                isRedundant = false;
+            }
+        }
+        
+        if(isRedundant) {
```

Comment:
Im not sure but i think we can let the user to update a post with the same details and it will simplify our design. you can disagree with me on that.